book.md

```diff
@@ -177,7 +177,146 @@ vulnerabilities, along with their mitigations, presented in a rough
+
+* writing a fixed value to an attacker-controlled address
+* writing to an address consisting of a fixed base and an attacker-controlled
+  offset limited to a specific range (e.g. a 32-bit offset)
+* writing to an attacker-controlled base address with a fixed offset
+
+Slide 11 of [@Miller2012] describes more properties of primitives that allow a
+more detailed classification.
+
+The most powerful version of a write primitive is an _arbitrary write_
+primitive, where both the address and the value are fully controlled by the
+attacker.
+
+A _read primitive_, respectively, gives the attacker read access to the
+victim's memory space. As for the write primitive, the address of the memory
+location accessed will be controlled by the attacker to some degree. One of the
+attacker's early goals will often be to achieve an _arbitrary read_ primitive,
+in which the address is fully controlled by the attacker.
+
+The effects of a write primitive are perhaps easier to understand, as it
+has obvious side-effects: a value is written to the victim program's memory.
+But how does an attacker get to observe the result of a read primitive?
+
+This basically depends on whether the attack is interactive or non-interactive,
+terminology which is taken from [@Hu2016].
+
+* In an _interactive attack_, the attacker gives malicious input to the victim
+  program, observes the result, and prepares the next input. Observing the
+  results is done through looking at the victim program's output, for example
+  looking at a network packet transmitted by the victim. For an attacker to
+  be able to interact with the victim through these primitives, the victim
+  process must stay alive between the handling of the malicious inputs, i.e.
+  it must not be restarted. Note that although the crafting of malicious
+  input could in theory be done by hand, it's usually done via an external
+  program that drives the exploit. An example of this type of attack can be
+  seen in [@Beer2020], which describes a zero-click radio proximity
+  exploit.
+* In a _non-interactive (one-shot) attack_, the attacker provides all malicious
+  input to the victim program at once. The malicious input triggers multiple
+  primitives one after the other, and the primitives are able to observe the
+  effects of the preceding operations through the victim program's state.
+  The input could be, for example, in the form of a JavaScript program
+  [@Groß2020], or a PDF file pretending to be a GIF [@Beer2021].
+
+How does an attacker achieve these kinds of primitives in the first place?
+The details for this vary, and in some cases it takes a combination of many
+techniques, some of which are out of scope for this book. But we will be
+seeing a few of them in this chapter, for example a stack overflow
+resulting in a (restricted) write primitive when the input size exceeds what
+the program expected.
+
+How are primitives chained to perform multiple reads/writes?
```

Comment:
I think this is the first place in the text that refers to chaining primitives?
If so, I think that maybe a paragraph is needed above this to first explain what chaining primitives is and why it's needed. Then it's easier for the reader to then read about the how.

Reply:
I'll rephrase this part, including using the word "combine" instead of chain, at least in the first instance, to make this more self-evident.